File: src/lib/api/skills.ts
import { invoke } from "@tauri-apps/api/core";

export type SkillAvailability = "Registered" | "DependenciesPending" | "Ready" | "Broken";
export interface SkillDependency { name: string; requirement?: string; kind: string; installed: boolean; detail?: string; }
export interface SkillMetadata { id: string; name: string; description: string; tags: string[]; triggers: string[]; capabilities: string[]; requires_execution: boolean; source_path: string; enabled: boolean; availability: SkillAvailability; resources: string[]; dependencies: SkillDependency[]; sandbox: { root: string; runtime: string; workspace: string; permission: string; network: boolean }; execution_mode: "Local" | "ReadOnly" | "Sandbox"; allow_execution: boolean; }
export interface InstallPlan { skill_id: string; commands: string[]; network: boolean; target: string; missing_system_dependencies: string[]; }
export interface TerminalResult { command: string; cwd: string; stdout: string; stderr: string; exit_code: number | null; }
export interface InstallSession { session_id: string; sandbox_root: string; workspace: string; network: boolean; created_at: string; }
export interface SkillCandidate { path: string; name: string; description: string; }
export type SkillSourceScope = "project" | "user" | "system" | "custom";
export interface SkillSourceRoot { scope: SkillSourceScope; label: string; path: string; exists: boolean; readable: boolean; }
export interface DiscoveredSkill { id: string; name: string; description: string; path: string; scope: SkillSourceScope; source_label: string; registered: boolean; conflict_rank: number; conflict: boolean; }

export const listSkills = () => invoke<SkillMetadata[]>("list_skills");
export const registerSkill = (path: string) => invoke<SkillMetadata>("register_skill", { path });
export const unregisterSkill = (id: string) => invoke<void>("unregister_skill", { id });
export const loadSkill = (id: string, resource?: string) => invoke<{ content: string }>("load_skill", { id, resource });
export const skillInstallPlan = (id: string) => invoke<InstallPlan>("skill_install_plan", { id });
export const installSkillDependencies = (id: string) => invoke<void>("install_skill_dependencies", { id });
export const runSkillTerminal = (id: string, command: string) => invoke<TerminalResult>("run_skill_terminal", { id, command });
export const setSkillExecution = (id: string, mode: "Local" | "ReadOnly" | "Sandbox", allowExecution: boolean) => invoke<SkillMetadata>("set_skill_execution", { id, mode, allowExecution });

export const createSkillInstallSession = () => invoke<InstallSession>("create_skill_install_session");
export const runSkillInstallTerminal = (sessionId: string, command: string) => invoke<TerminalResult>("run_skill_install_terminal", { sessionId, command });
export const detectInstalledSkills = (sessionId: string) => invoke<SkillCandidate[]>("detect_installed_skills", { sessionId });
export const registerDetectedSkill = (sessionId: string, path: string) => invoke<SkillMetadata>("register_detected_skill", { sessionId, path });
export const cancelSkillInstallSession = (sessionId: string) => invoke<void>("cancel_skill_install_session", { sessionId });
export const discoverSkills = (workspacePath?: string, customRoots: string[] = []) => invoke<DiscoveredSkill[]>("discover_skills", { workspacePath, customRoots });
export const listSkillSourceRoots = (workspacePath?: string, customRoots: string[] = []) => invoke<SkillSourceRoot[]>("list_skill_source_roots", { workspacePath, customRoots });
export const registerDiscoveredSkill = (path: string) => invoke<SkillMetadata>("register_discovered_skill", { path });
export const registerDiscoveredSkillsBatch = (paths: string[]) => invoke<SkillMetadata[]>("register_discovered_skills_batch", { paths });

File: src-tauri/src/commands/skills.rs
use crate::error::{AppError, AppResult};
use crate::skills::{
    self,
    model::{InstallPlan, InstallSession, SkillCandidate, SkillMetadata},
};
use std::path::PathBuf;

#[tauri::command]
pub fn list_skills() -> AppResult<Vec<SkillMetadata>> {
    skills::list().map_err(AppError::Message)
}

#[tauri::command]
pub fn discover_skills(
    workspace_path: Option<String>,
    custom_roots: Vec<String>,
) -> AppResult<Vec<crate::skills::discovery::DiscoveredSkill>> {
    let workspace = workspace_path.map(PathBuf::from);
    let roots = custom_roots.into_iter().map(PathBuf::from).collect();
    skills::discover(workspace, roots).map_err(AppError::Message)
}

#[tauri::command]
pub fn list_skill_source_roots(
    workspace_path: Option<String>,
    custom_roots: Vec<String>,
) -> AppResult<Vec<crate::skills::discovery::SkillSourceRoot>> {
    let workspace = workspace_path.map(PathBuf::from);
    let roots = custom_roots
        .into_iter()
        .map(PathBuf::from)
        .collect::<Vec<_>>();
    Ok(crate::skills::discovery::source_roots(
        workspace.as_deref(),
        &roots,
    ))
}

#[tauri::command]
pub fn register_discovered_skill(path: String) -> AppResult<SkillMetadata> {
    skills::register(PathBuf::from(path)).map_err(AppError::Message)
}

#[tauri::command]
pub fn register_discovered_skills_batch(paths: Vec<String>) -> AppResult<Vec<SkillMetadata>> {
    let path_bufs = paths.into_iter().map(PathBuf::from).collect();
    skills::register_batch(path_bufs).map_err(AppError::Message)
}

#[tauri::command]
pub fn register_skill(path: String) -> AppResult<SkillMetadata> {
    skills::register(PathBuf::from(path)).map_err(AppError::Message)
}

#[tauri::command]
pub fn unregister_skill(id: String) -> AppResult<()> {
    skills::unregister(&id).map_err(AppError::Message)
}

#[tauri::command]
pub fn set_skill_execution(
    id: String,
    mode: skills::model::SkillExecutionMode,
    allow_execution: bool,
) -> AppResult<SkillMetadata> {
    let mut store = skills::storage::load().map_err(|e| AppError::Message(e.to_string()))?;
    let registration = store
        .registrations
        .iter_mut()
        .find(|item| item.id == id)
        .ok_or_else(|| AppError::Message("skill not found".into()))?;
    registration.execution_mode = mode;
    registration.allow_execution = allow_execution;
    let updated = registration.clone();
    skills::storage::save(&store).map_err(|e| AppError::Message(e.to_string()))?;
    skills::metadata(&updated).map_err(AppError::Message)
}

#[tauri::command]
pub fn load_skill(id: String, resource: Option<String>) -> AppResult<serde_json::Value> {
    skills::load_content(&id, resource.as_deref()).map_err(AppError::Message)
}

#[tauri::command]
pub fn run_skill_terminal(
    id: String,
    command: String,
) -> AppResult<crate::skills::terminal::TerminalResult> {
    skills::terminal::run(&id, &command).map_err(AppError::Message)
}

#[tauri::command]
pub fn install_skill_dependencies(id: String) -> AppResult<()> {
    skills::installer::install(&id).map_err(AppError::Message)
}

#[tauri::command]
pub fn skill_install_plan(id: String) -> AppResult<InstallPlan> {
    let meta = skills::list()
        .map_err(AppError::Message)?
        .into_iter()
        .find(|item| item.id == id)
        .ok_or_else(|| AppError::Message("skill not found".into()))?;
    Ok(InstallPlan { skill_id: id, commands: meta.dependencies.iter().filter_map(|dep| match dep.kind { skills::model::DependencyKind::Node => Some("npm install --prefix runtime".into()), skills::model::DependencyKind::Python => Some("python -m venv runtime/python && runtime/python/bin/pip install -r requirements.txt".into()), _ => None }).collect(), network: !meta.dependencies.is_empty(), target: meta.sandbox.runtime, missing_system_dependencies: Vec::new() })
}

#[tauri::command]
pub fn create_skill_install_session() -> AppResult<InstallSession> {
    skills::install_session::create_session(true).map_err(AppError::from)
}

#[tauri::command]
pub fn run_skill_install_terminal(
    session_id: String,
    command: String,
) -> AppResult<crate::skills::terminal::TerminalResult> {
    skills::install_session::run_terminal(&session_id, &command).map_err(AppError::Message)
}

#[tauri::command]
pub fn detect_installed_skills(session_id: String) -> AppResult<Vec<SkillCandidate>> {
    skills::install_session::detect_candidates(&session_id).map_err(AppError::Message)
}

#[tauri::command]
pub fn register_detected_skill(session_id: String, path: String) -> AppResult<SkillMetadata> {
    skills::install_session::register_from_session(&session_id, &path).map_err(AppError::Message)
}

#[tauri::command]
pub fn cancel_skill_install_session(session_id: String) -> AppResult<()> {
    skills::install_session::cleanup_session(&session_id).map_err(AppError::Message)
}
